stuck in endless bootloop / maybe camera module broken?

Hi there,

first of all thank you for this awesome project!

My esp32cam device won't boot anymore and is stuck in an endless bootloop.

This is it's last response:
```
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.
Core 0 register dump:
PC      : 0x400eef32  PS      : 0x00060330  A0      : 0x8018c841  A1      : 0x3fff7c00  
A2      : 0x00000000  A3      : 0xfffffffe  A4      : 0xffffff9c  A5      : 0xffffff9c  
A6      : 0x3ffb7c50  A7      : 0x00000001  A8      : 0xffffff9d  A9      : 0x3fff7be0  
A10     : 0x00000000  A11     : 0x00000000  A12     : 0x3ffd3853  A13     : 0x00000000  
A14     : 0x00000000  A15     : 0xff000000  SAR     : 0x00000018  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000048  LBEG    : 0x4008a3cc  LEND    : 0x4008a3e8  LCOUNT  : 0xffffffff  

ELF file SHA256: 0dbda5e7af80986c

Backtrace: 0x400eef2f:0x3fff7c00 0x4018c83e:0x3fff7c20 0x4018b335:0x3fff7ce0 0x400f1066:0x3fff8160 0x400f126b:0x3fff81a0

Rebooting...
```

It seems to me that my camera module is broken or can't be initialized what might cause this loop:
```
 [0;32mI (4019) gpio: GPIO[32]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0  [0m
 [0;31mE (4089) camera: Detected camera not supported. [0m
 [0;31mE (4089) camera: Camera probe failed with error 0x20004 [0m
 [0;31mE (4089) server_part_camera: Camera Init Failed [0m
```

Could there by any other cause? And if not, would it be possible to get an error message on the dashboard instead by handling that exception?


Best regards
Sven

<details>
<summary>full boot log</summary>

```
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4
load:0x3fff0034,len:7248
ho 0 tail 12 room 4
load:0x40078000,len:13856
load:0x40080400,len:4052
entry 0x40080688
 [0;32mI (31) boot: ESP-IDF 3.40100.200827 2nd stage bootloader [0m
 [0;32mI (31) boot: compile time 07:07:29 [0m
 [0;32mI (31) boot: chip revision: 1 [0m
 [0;32mI (34) boot_comm: chip revision: 1, min. bootloader chip revision: 0 [0m
 [0;32mI (41) boot.esp32: SPI Speed      : 40MHz [0m
 [0;32mI (46) boot.esp32: SPI Mode       : DIO [0m
 [0;32mI (51) boot.esp32: SPI Flash Size : 4MB [0m
 [0;32mI (55) boot: Enabling RNG early entropy source... [0m
 [0;32mI (61) boot: Partition Table: [0m
 [0;32mI (64) boot: ## Label            Usage          Type ST Offset   Length [0m
 [0;32mI (71) boot:  0 nvs              WiFi data        01 02 00009000 00004000 [0m
 [0;32mI (79) boot:  1 otadata          OTA data         01 00 0000d000 00002000 [0m
 [0;32mI (86) boot:  2 phy_init         RF data          01 01 0000f000 00001000 [0m
 [0;32mI (94) boot:  3 ota_0            OTA app          00 10 00010000 001db000 [0m
 [0;32mI (101) boot:  4 ota_1            OTA app          00 11 001f0000 001db000 [0m
 [0;32mI (109) boot: End of partition table [0m
 [0;32mI (113) boot_comm: chip revision: 1, min. application chip revision: 0 [0m
 [0;32mI (120) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x50f90 (331664) map [0m
 [0;32mI (273) esp_image: segment 1: paddr=0x00060fb8 vaddr=0x3ffb0000 size=0x05050 ( 20560) load [0m
 [0;32mI (283) esp_image: segment 2: paddr=0x00066010 vaddr=0x40080000 size=0x00404 (  1028) load [0m
 [0;32mI (284) esp_image: segment 3: paddr=0x0006641c vaddr=0x40080404 size=0x09bfc ( 39932) load [0m
 [0;32mI (310) esp_image: segment 4: paddr=0x00070020 vaddr=0x400d0020 size=0x122edc (1191644) map [0m
 [0;32mI (826) esp_image: segment 5: paddr=0x00192f04 vaddr=0x4008a000 size=0x12bb4 ( 76724) load [0m
 [0;32mI (880) boot: Loaded app from partition at offset 0x10000 [0m
 [0;32mI (880) boot: Disabling RNG early entropy source... [0m
 [0;32mI (881) psram: This chip is ESP32-D0WD [0m
 [0;32mI (885) spiram: Found 64MBit SPI RAM device [0m
 [0;32mI (890) spiram: SPI RAM mode: flash 40m sram 40m [0m
 [0;32mI (895) spiram: PSRAM initialized, cache is in low/high (2-core) mode. [0m
 [0;32mI (902) cpu_start: Pro cpu up. [0m
 [0;32mI (906) cpu_start: Application information: [0m
 [0;32mI (911) cpu_start: Project name:     esp32cam-server-only [0m
 [0;32mI (917) cpu_start: App version:      0.0.9.3 [0m
 [0;32mI (922) cpu_start: Compile time:     Nov 16 2021 07:06:13 [0m
 [0;32mI (928) cpu_start: ELF file SHA256:  0dbda5e7af80986c... [0m
 [0;32mI (934) cpu_start: ESP-IDF:          3.40100.200827 [0m
 [0;32mI (940) cpu_start: Starting app cpu, entry point is 0x40081b00 [0m
 [0;32mI (931) cpu_start: App cpu up. [0m
 [0;32mI (1837) spiram: SPI SRAM memory test OK [0m
 [0;32mI (1838) heap_init: Initializing. RAM available for dynamic allocation: [0m
 [0;32mI (1838) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM [0m
 [0;32mI (1844) heap_init: At 3FFC2A10 len 0001D5F0 (117 KiB): DRAM [0m
 [0;32mI (1851) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM [0m
 [0;32mI (1857) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM [0m
 [0;32mI (1864) heap_init: At 4009CBB4 len 0000344C (13 KiB): IRAM [0m
 [0;32mI (1870) cpu_start: Pro cpu start user code [0m
 [0;32mI (1875) spiram: Adding pool of 4096K of external SPI memory to heap allocator [0m
 [0;32mI (1900) spi_flash: detected chip: generic [0m
 [0;32mI (1901) spi_flash: flash io: dio [0m
 [0;32mI (1901) cpu_start: Starting scheduler on PRO CPU. [0m
 [0;32mI (0) cpu_start: Starting scheduler on APP CPU. [0m
 [0;32mI (1909) spiram: Reserving pool of 32K of internal memory for DMA/internal allocations [0m
Do Reset Camera
 [0;32mI (1919) gpio: GPIO[32]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0  [0m
Init Camera
 [0;32mI (3929) gpio: GPIO[5]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3929) gpio: GPIO[18]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3929) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3939) gpio: GPIO[21]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3949) gpio: GPIO[22]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3959) gpio: GPIO[23]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3969) gpio: GPIO[25]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3979) gpio: GPIO[34]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3989) gpio: GPIO[35]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (3999) gpio: GPIO[36]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (4009) gpio: GPIO[39]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0  [0m
 [0;32mI (4019) sccb: pin_sda 26 pin_scl 27
 [0m
 [0;32mI (4019) gpio: GPIO[32]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0  [0m
 [0;31mE (4089) camera: Detected camera not supported. [0m
 [0;31mE (4089) camera: Camera probe failed with error 0x20004 [0m
 [0;31mE (4089) server_part_camera: Camera Init Failed [0m
 [0;32mI (4139) main: Using SDMMC peripheral [0m
Name: SD08G
Type: SDHC/SDXC
Speed: 20 MHz
Size: 7600MB
 [0;32mI (4289) server_ota: Start CheckOTAUpdateCheck ... [0m
Start CheckOTAUpdateCheck ...
 [0;32mI (4299) server_ota: SHA-256 for the partition table: : 9d4a2809ae90fb20709a8ca3dd3ac3462e1e73286bcd31b689060c91de06fa3a [0m
 [0;32mI (4359) server_ota: SHA-256 for bootloader: : 2f4c792de3ca36713d14c6b1e98c2601ad2b22206ee169706b4dcf9e39c7c421 [0m
 [0;32mI (4999) server_ota: SHA-256 for current firmware: : fd924f0025e85034e06f0e4e65c62dd323f6af611ccfc42f24d89530b300852d [0m
CheckOTAUpdate Partition: ESP_OK
open file /sdcard/wlan.ini in mode r
file loaded

WLan: WifiSSID, WifiPassword
Hostename: gasmeter
I (5059) wifi:wifi driver task: 3ffd36a4, prio:23, stack:6656, core=0
 [0;32mI (5059) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE [0m
 [0;32mI (5059) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE [0m
I (5089) wifi:wifi firmware version: 3ea4c76
I (5089) wifi:config NVS flash: enabled
I (5089) wifi:config nano formating: disabled
I (5089) wifi:Init dynamic tx buffer num: 32
I (5099) wifi:Init data frame dynamic rx buffer num: 32
I (5099) wifi:Init management frame dynamic rx buffer num: 32
I (5109) wifi:Init management short buffer num: 32
I (5109) wifi:Init static tx buffer num: 16
I (5119) wifi:Init static rx buffer size: 1600
I (5119) wifi:Init static rx buffer num: 10
I (5119) wifi:Init dynamic rx buffer num: 32
 [0;32mI (5229) phy: phy_version: 4180, cb3948e, Sep 12 2019, 16:39:13, 0, 0 [0m
I (5229) wifi:mode : sta (7c:9e:bd:60:fd:d8)
 [0;32mI (5229) BLINK: Blinken - start [0m
 [0;32mI (5239) wifi station: wifi_init_sta finished. [0m
 [0;32mI (5439) BLINK: Blinken - done [0m
I (5599) wifi:new:<13,2>, old:<1,0>, ap:<255,255>, sta:<13,2>, prof:1
I (5599) wifi:state: init -> auth (b0)
I (5609) wifi:state: auth -> assoc (0)
I (5629) wifi:state: assoc -> run (10)
I (5669) wifi:connected with WifiSSID, aid = 1, channel 13, 40D, bssid = 3e:94:ed:c5:8e:12
I (5669) wifi:security: WPA2-PSK, phy: bgn, rssi: -45
I (5679) wifi:pm start, type: 1

I (5769) wifi:AP's beacon interval = 204800 us, DTIM period = 4
 [0;32mI (6539) esp_netif_handlers: sta ip: 192.168.1.80, mask: 255.255.255.0, gw: 192.168.1.1 [0m
 [0;32mI (6539) wifi station: got ip:192.168.1.80 [0m
 [0;32mI (6539) BLINK: Blinken - start [0m
 [0;32mI (6539) wifi station: connected to ap SSID:WifiSSID password:WifiPassword [0m
main: sleep for : 200ms
 [0;32mI (8549) sntp: Time is not set yet. Connecting to WiFi and getting time over NTP. [0m
 [0;32mI (8549) sntp: Initializing SNTP [0m
 [0;32mI (8549) sntp: Waiting for system time to be set... (1/10) [0m
 [0;32mI (10549) sntp: Waiting for system time to be set... (2/10) [0m
 [0;32mI (12549) sntp: Waiting for system time to be set... (3/10) [0m
TimeZone set to CET-1CEST,M3.5.0,M10.5.0/3
Logfile opened: /sdcard/log/message/log_2021-11-20.txt
 [0;32mI (14569) sntp: The current date/time in Berlin is: Sat Nov 20 17:02:58 2021 [0m
 [0;32mI (14569) sntp: The current date/time in Berlin is: 2021-11-20_17:02 [0m
timeist 20211120-170258
Logfile opened: /sdcard/log/message/log_2021-11-20.txt
Logfile opened: /sdcard/log/message/log_2021-11-20.txt
Logfile opened: /sdcard/log/message/log_2021-11-20.txt
time 20211120-170258
main: sleep for : 200ms
 [0;32mI (16549) BLINK: Blinken - done [0m
 [0;32mI (16639) server-main: Starting server on port: '80' [0m
 [0;32mI (16639) server-main: Registering URI handlers [0m
 [0;32mI (16639) server_tflite: server_part_camera - Registering URI handlers [0m
 [0;32mI (16639) server_ota: server_ota - Registering URI handlers [0m
 [0;32mI (16649) server_GPIO: start GpioHandler [0m
 [0;32mI (16649) server_GPIO: register GPIO Uri [0m
 [0;32mI (16659) server_GPIO: server_GPIO - Registering URI handlers [0m
vor reg server main
vor dotautostart
task_autodoFlow: start

open file /sdcard/config/config.ini in mode r
[MakeImage]
Start ReadParameter ([MakeImage]
)
;LogImageLocation = /log/source
WaitBeforeTakingPicture = 5
;LogfileRetentionInDays = 15
Brightness = -2
;Contrast = 0
;Saturation = 0
ImageQuality = 5
ImageSize = VGA
FixedExposure = false

[Alignment]
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.
Core 0 register dump:
PC      : 0x400eef32  PS      : 0x00060330  A0      : 0x8018c841  A1      : 0x3fff7c00  
A2      : 0x00000000  A3      : 0xfffffffe  A4      : 0xffffff9c  A5      : 0xffffff9c  
A6      : 0x3ffb7c50  A7      : 0x00000001  A8      : 0xffffff9d  A9      : 0x3fff7be0  
A10     : 0x00000000  A11     : 0x00000000  A12     : 0x3ffd3853  A13     : 0x00000000  
A14     : 0x00000000  A15     : 0xff000000  SAR     : 0x00000018  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000048  LBEG    : 0x4008a3cc  LEND    : 0x4008a3e8  LCOUNT  : 0xffffffff  

ELF file SHA256: 0dbda5e7af80986c

Backtrace: 0x400eef2f:0x3fff7c00 0x4018c83e:0x3fff7c20 0x4018b335:0x3fff7ce0 0x400f1066:0x3fff8160 0x400f126b:0x3fff81a0

Rebooting...
```
</details>

Hi @jomjol,

I'm struggeling to compile the sources using platformio. Do you have any suggestions what might be missing?

Best regards,
ottk3

PS C:\Users\username\sources\AI-on-the-edge-device\code> platformio run --environment esp32cam
```
Processing esp32cam (platform: espressif32; board: esp32cam; framework: espidf)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32cam.html
PLATFORM: Espressif 32 (3.4.0) > AI Thinker ESP32-CAM
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-espidf 3.40301.0 (4.3.1)
 - tool-cmake 3.16.4
 - tool-esptoolpy 1.30100.210531 (3.1.0)
 - tool-idf 1.0.1
 - tool-mconf 1.4060000.20190628 (406.0.0)
 - tool-ninja 1.9.0
 - toolchain-esp32ulp 1.22851.191205 (2.28.51)
 - toolchain-riscv32-esp 8.4.0+2021r1
 - toolchain-xtensa-esp32 8.4.0+2021r1
 - toolchain-xtensa-esp32s2 8.4.0+2021r1
Reading CMake configuration...
Library Manager: Installing jomjol_configfile
Warning! Could not find the package with 'jomjol_configfile' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_helper
Warning! Could not find the package with 'jomjol_helper' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_wlan
Warning! Could not find the package with 'jomjol_wlan' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_image_proc
Warning! Could not find the package with 'jomjol_image_proc' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_controlcamera
Warning! Could not find the package with 'jomjol_controlcamera' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_flowcontroll
Warning! Could not find the package with 'jomjol_flowcontroll' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_tfliteclass
Warning! Could not find the package with 'jomjol_tfliteclass' requirements for your system 'windows_amd64'
Library Manager: Installing tfmicro
Warning! Could not find the package with 'tfmicro' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_fileserver_ota
Warning! Could not find the package with 'jomjol_fileserver_ota' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_time_sntp
Warning! Could not find the package with 'jomjol_time_sntp' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_logfile
Warning! Could not find the package with 'jomjol_logfile' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_mqtt
Warning! Could not find the package with 'jomjol_mqtt' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_controlGPIO
Warning! Could not find the package with 'jomjol_controlGPIO' requirements for your system 'windows_amd64'
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Library Manager: Installing jomjol_configfile
Warning! Could not find the package with 'jomjol_configfile' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_helper
Warning! Could not find the package with 'jomjol_helper' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_wlan
Warning! Could not find the package with 'jomjol_wlan' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_image_proc
Warning! Could not find the package with 'jomjol_image_proc' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_controlcamera
Warning! Could not find the package with 'jomjol_controlcamera' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_flowcontroll
Warning! Could not find the package with 'jomjol_flowcontroll' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_tfliteclass
Warning! Could not find the package with 'jomjol_tfliteclass' requirements for your system 'windows_amd64'
Library Manager: Installing tfmicro
Warning! Could not find the package with 'tfmicro' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_fileserver_ota
Warning! Could not find the package with 'jomjol_fileserver_ota' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_time_sntp
Warning! Could not find the package with 'jomjol_time_sntp' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_logfile
Warning! Could not find the package with 'jomjol_logfile' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_mqtt
Warning! Could not find the package with 'jomjol_mqtt' requirements for your system 'windows_amd64'
Library Manager: Installing jomjol_controlGPIO
Warning! Could not find the package with 'jomjol_controlGPIO' requirements for your system 'windows_amd64'
Found 0 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Archiving .pio\build\esp32cam\esp-idf\jomjol_controlGPIO\libjomjol_controlGPIO.a
Archiving .pio\build\esp32cam\esp-idf\jomjol_controlcamera\libjomjol_controlcamera.a
Compiling .pio\build\esp32cam\components\jomjol_fileserver_ota\server_file.o
Compiling .pio\build\esp32cam\components\jomjol_fileserver_ota\server_help.o
Compiling .pio\build\esp32cam\components\jomjol_flowcontroll\ClassFlowControll.o
Compiling .pio\build\esp32cam\components\jomjol_flowcontroll\ClassFlowImage.o
Compiling .pio\build\esp32cam\components\jomjol_helper\Helper.o
Compiling .pio\build\esp32cam\components\jomjol_image_proc\CFindTemplate.o
Compiling .pio\build\esp32cam\components\jomjol_image_proc\CImageBasis.o
Compiling .pio\build\esp32cam\components\jomjol_image_proc\CRotateImage.o
Compiling .pio\build\esp32cam\components\jomjol_image_proc\make_stb.o
Compiling .pio\build\esp32cam\components\jomjol_logfile\ClassLogFile.o
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:53:6: error: conflicting declaration of 'DIR* opendir(const char*)' with 'C' linkage
 DIR *opendir(const char *);
      ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:53:6: note: previous declaration with 'C++' linkage
 DIR* opendir(const char* name);
      ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:56:3: error: conflicting declaration of 'dirent* readdir(DIR*)' with 'C' linkage
   readdir(DIR *);
   ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:54:16: note: previous declaration with 'C++' linkage
 struct dirent* readdir(DIR* pdir);
                ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:58:6: error: conflicting declaration of 'int readdir_r(DIR*, dirent*, dirent**)' with 'C' linkage
 int  readdir_r(DIR *__restrict, struct dirent *__restrict,
      ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:59:5: note: previous declaration with 'C++' linkage
 int readdir_r(DIR* pdir, struct dirent* entry, struct dirent** out_dirent);
     ^~~~~~~~~
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:61:7: error: conflicting declaration of 'void rewinddir(DIR*)' with 'C' linkage
 void  rewinddir(DIR *);
       ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:57:6: note: previous declaration with 'C++' linkage
 void rewinddir(DIR* pdir);
      ^~~~~~~~~
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:72:7: error: conflicting declaration of 'void seekdir(DIR*, long int)' with 'C' linkage
 void  seekdir(DIR *, long);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:56:6: note: previous declaration with 'C++' linkage
 void seekdir(DIR* pdir, long loc);
      ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:73:7: error: conflicting declaration of 'long int telldir(DIR*)' with 'C' linkage
 long  telldir(DIR *);
Compiling .pio\build\esp32cam\components\jomjol_mqtt\interface_mqtt.o
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:55:6: note: previous declaration with 'C++' linkage
 long telldir(DIR* pdir);
      ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_help.cpp:8:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:76:6: error: conflicting declaration of 'int closedir(DIR*)' with 'C' linkage
 int  closedir(DIR *);
      ^~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_help.cpp:8:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:58:5: note: previous declaration with 'C++' linkage
 int closedir(DIR* pdir);
     ^~~~~~~~
Compiling .pio\build\esp32cam\components\jomjol_tfliteclass\CTfLiteClass.o
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:53:6: error: conflicting declaration of 'DIR* opendir(const char*)' with 'C' linkage
 DIR *opendir(const char *);
      ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:53:6: note: previous declaration with 'C++' linkage
 DIR* opendir(const char* name);
      ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:56:3: error: conflicting declaration of 'dirent* readdir(DIR*)' with 'C' linkage
   readdir(DIR *);
   ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:54:16: note: previous declaration with 'C++' linkage
 struct dirent* readdir(DIR* pdir);
                ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:58:6: error: conflicting declaration of 'int readdir_r(DIR*, dirent*, dirent**)' with 'C' linkage
 int  readdir_r(DIR *__restrict, struct dirent *__restrict,
      ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:59:5: note: previous declaration with 'C++' linkage
 int readdir_r(DIR* pdir, struct dirent* entry, struct dirent** out_dirent);
     ^~~~~~~~~
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:61:7: error: conflicting declaration of 'void rewinddir(DIR*)' with 'C' linkage
 void  rewinddir(DIR *);
       ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:57:6: note: previous declaration with 'C++' linkage
 void rewinddir(DIR* pdir);
      ^~~~~~~~~
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:72:7: error: conflicting declaration of 'void seekdir(DIR*, long int)' with 'C' linkage
 void  seekdir(DIR *, long);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:56:6: note: previous declaration with 'C++' linkage
 void seekdir(DIR* pdir, long loc);
      ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:73:7: error: conflicting declaration of 'long int telldir(DIR*)' with 'C' linkage
 long  telldir(DIR *);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:55:6: note: previous declaration with 'C++' linkage
 long telldir(DIR* pdir);
      ^~~~~~~
In file included from components\jomjol_fileserver_ota\server_file.cpp:20:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:76:6: error: conflicting declaration of 'int closedir(DIR*)' with 'C' linkage
 int  closedir(DIR *);
      ^~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_fileserver_ota\server_file.cpp:20:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:58:5: note: previous declaration with 'C++' linkage
 int closedir(DIR* pdir);
     ^~~~~~~~
Compiling .pio\build\esp32cam\components\jomjol_tfliteclass\server_tflite.o
Compiling .pio\build\esp32cam\components\jomjol_time_sntp\time_sntp.o
*** [.pio\build\esp32cam\components\jomjol_fileserver_ota\server_help.o] Error 1
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:53:6: error: conflicting declaration of 'DIR* opendir(const char*)' with 'C' linkage
 DIR *opendir(const char *);
      ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:53:6: note: previous declaration with 'C++' linkage
 DIR* opendir(const char* name);
      ^~~~~~~
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:56:3: error: conflicting declaration of 'dirent* readdir(DIR*)' with 'C' linkage
   readdir(DIR *);
   ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:54:16: note: previous declaration with 'C++' linkage
 struct dirent* readdir(DIR* pdir);
                ^~~~~~~
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:58:6: error: conflicting declaration of 'int readdir_r(DIR*, dirent*, dirent**)' with 'C' linkage
 int  readdir_r(DIR *__restrict, struct dirent *__restrict,
      ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:59:5: note: previous declaration with 'C++' linkage
 int readdir_r(DIR* pdir, struct dirent* entry, struct dirent** out_dirent);
     ^~~~~~~~~
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:61:7: error: conflicting declaration of 'void rewinddir(DIR*)' with 'C' linkage
 void  rewinddir(DIR *);
       ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:57:6: note: previous declaration with 'C++' linkage
 void rewinddir(DIR* pdir);
      ^~~~~~~~~
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:72:7: error: conflicting declaration of 'void seekdir(DIR*, long int)' with 'C' linkage
 void  seekdir(DIR *, long);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:56:6: note: previous declaration with 'C++' linkage
 void seekdir(DIR* pdir, long loc);
      ^~~~~~~
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:73:7: error: conflicting declaration of 'long int telldir(DIR*)' with 'C' linkage
 long  telldir(DIR *);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:55:6: note: previous declaration with 'C++' linkage
 long telldir(DIR* pdir);
      ^~~~~~~
In file included from components\jomjol_helper\Helper.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:76:6: error: conflicting declaration of 'int closedir(DIR*)' with 'C' linkage
 int  closedir(DIR *);
      ^~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_helper\Helper.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:58:5: note: previous declaration with 'C++' linkage
 int closedir(DIR* pdir);
     ^~~~~~~~
*** [.pio\build\esp32cam\components\jomjol_fileserver_ota\server_file.o] Error 1
*** [.pio\build\esp32cam\components\jomjol_helper\Helper.o] Error 1
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:53:6: error: conflicting declaration of 'DIR* opendir(const char*)' with 'C' linkage
 DIR *opendir(const char *);
      ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:53:6: note: previous declaration with 'C++' linkage
 DIR* opendir(const char* name);
      ^~~~~~~
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:56:3: error: conflicting declaration of 'dirent* readdir(DIR*)' with 'C' linkage
   readdir(DIR *);
   ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:54:16: note: previous declaration with 'C++' linkage
 struct dirent* readdir(DIR* pdir);
                ^~~~~~~
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:58:6: error: conflicting declaration of 'int readdir_r(DIR*, dirent*, dirent**)' with 'C' linkage
 int  readdir_r(DIR *__restrict, struct dirent *__restrict,
      ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:59:5: note: previous declaration with 'C++' linkage
 int readdir_r(DIR* pdir, struct dirent* entry, struct dirent** out_dirent);
     ^~~~~~~~~
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:61:7: error: conflicting declaration of 'void rewinddir(DIR*)' with 'C' linkage
 void  rewinddir(DIR *);
       ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:57:6: note: previous declaration with 'C++' linkage
 void rewinddir(DIR* pdir);
      ^~~~~~~~~
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:72:7: error: conflicting declaration of 'void seekdir(DIR*, long int)' with 'C' linkage
 void  seekdir(DIR *, long);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:56:6: note: previous declaration with 'C++' linkage
 void seekdir(DIR* pdir, long loc);
      ^~~~~~~
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:73:7: error: conflicting declaration of 'long int telldir(DIR*)' with 'C' linkage
 long  telldir(DIR *);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:55:6: note: previous declaration with 'C++' linkage
 long telldir(DIR* pdir);
      ^~~~~~~
In file included from components\jomjol_logfile\ClassLogFile.cpp:6:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:76:6: error: conflicting declaration of 'int closedir(DIR*)' with 'C' linkage
 int  closedir(DIR *);
      ^~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_logfile\ClassLogFile.cpp:6:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:58:5: note: previous declaration with 'C++' linkage
 int closedir(DIR* pdir);
     ^~~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:53:6: error: conflicting declaration of 'DIR* opendir(const char*)' with 'C' linkage
 DIR *opendir(const char *);
      ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:53:6: note: previous declaration with 'C++' linkage
 DIR* opendir(const char* name);
      ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:56:3: error: conflicting declaration of 'dirent* readdir(DIR*)' with 'C' linkage
   readdir(DIR *);
   ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:54:16: note: previous declaration with 'C++' linkage
 struct dirent* readdir(DIR* pdir);
                ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:58:6: error: conflicting declaration of 'int readdir_r(DIR*, dirent*, dirent**)' with 'C' linkage
 int  readdir_r(DIR *__restrict, struct dirent *__restrict,
      ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:59:5: note: previous declaration with 'C++' linkage
 int readdir_r(DIR* pdir, struct dirent* entry, struct dirent** out_dirent);
     ^~~~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:61:7: error: conflicting declaration of 'void rewinddir(DIR*)' with 'C' linkage
 void  rewinddir(DIR *);
       ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:57:6: note: previous declaration with 'C++' linkage
 void rewinddir(DIR* pdir);
      ^~~~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:72:7: error: conflicting declaration of 'void seekdir(DIR*, long int)' with 'C' linkage
 void  seekdir(DIR *, long);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:56:6: note: previous declaration with 'C++' linkage
 void seekdir(DIR* pdir, long loc);
      ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:73:7: error: conflicting declaration of 'long int telldir(DIR*)' with 'C' linkage
 long  telldir(DIR *);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:55:6: note: previous declaration with 'C++' linkage
 long telldir(DIR* pdir);
      ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:76:6: error: conflicting declaration of 'int closedir(DIR*)' with 'C' linkage
 int  closedir(DIR *);
      ^~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowImage.cpp:5:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:58:5: note: previous declaration with 'C++' linkage
 int closedir(DIR* pdir);
     ^~~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:53:6: error: conflicting declaration of 'DIR* opendir(const char*)' with 'C' linkage
 DIR *opendir(const char *);
      ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:53:6: note: previous declaration with 'C++' linkage
 DIR* opendir(const char* name);
      ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:56:3: error: conflicting declaration of 'dirent* readdir(DIR*)' with 'C' linkage
   readdir(DIR *);
   ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:54:16: note: previous declaration with 'C++' linkage
 struct dirent* readdir(DIR* pdir);
                ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:58:6: error: conflicting declaration of 'int readdir_r(DIR*, dirent*, dirent**)' with 'C' linkage
 int  readdir_r(DIR *__restrict, struct dirent *__restrict,
      ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:59:5: note: previous declaration with 'C++' linkage
 int readdir_r(DIR* pdir, struct dirent* entry, struct dirent** out_dirent);
     ^~~~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:61:7: error: conflicting declaration of 'void rewinddir(DIR*)' with 'C' linkage
 void  rewinddir(DIR *);
       ^~~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:57:6: note: previous declaration with 'C++' linkage
 void rewinddir(DIR* pdir);
      ^~~~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:72:7: error: conflicting declaration of 'void seekdir(DIR*, long int)' with 'C' linkage
 void  seekdir(DIR *, long);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:56:6: note: previous declaration with 'C++' linkage
 void seekdir(DIR* pdir, long loc);
      ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:73:7: error: conflicting declaration of 'long int telldir(DIR*)' with 'C' linkage
 long  telldir(DIR *);
       ^~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:55:6: note: previous declaration with 'C++' linkage
 long telldir(DIR* pdir);
      ^~~~~~~
In file included from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:76:6: error: conflicting declaration of 'int closedir(DIR*)' with 'C' linkage
 int  closedir(DIR *);
      ^~~~~~~~
In file included from c:\users\username\.platformio\packages\toolchain-xtensa-esp32\xtensa-esp32-elf\sys-include\dirent.h:39,
                 from components\jomjol_flowcontroll\ClassFlowControll.cpp:9:
C:\Users\username\.platformio\packages\framework-espidf\components\newlib\platform_include/sys/dirent.h:58:5: note: previous declaration with 'C++' linkage
 int closedir(DIR* pdir);
     ^~~~~~~~
*** [.pio\build\esp32cam\components\jomjol_flowcontroll\ClassFlowImage.o] Error 1
*** [.pio\build\esp32cam\components\jomjol_logfile\ClassLogFile.o] Error 1
components\jomjol_mqtt\interface_mqtt.cpp: In function 'void MQTTInit(std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, int)':
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::event_handle' [-Wmissing-field-initializers]
     };
     ^
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::event_loop_handle' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::host' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::port' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::username' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::password' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::lwt_qos' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::disable_clean_session' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::disable_auto_reconnect' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::user_context' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::task_prio' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::task_stack' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::buffer_size' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::cert_pem' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::cert_len' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::client_cert_pem' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::client_cert_len' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::client_key_pem' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::client_key_len' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::transport' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::refresh_connection_after_ms' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::psk_hint_key' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::use_global_ca_store' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::crt_bundle_attach' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::reconnect_timeout_ms' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::alpn_protos' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::clientkey_password' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::clientkey_password_len' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::protocol_ver' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::out_buffer_size' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::skip_cert_common_name_check' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::use_secure_element' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::ds_data' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::network_timeout_ms' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::disable_keepalive' [-Wmissing-field-initializers]
components\jomjol_mqtt\interface_mqtt.cpp:105:5: warning: missing initializer for member 'esp_mqtt_client_config_t::path' [-Wmissing-field-initializers]
*** [.pio\build\esp32cam\components\jomjol_flowcontroll\ClassFlowControll.o] Error 1
================================================================================================== [FAILED] Took 12.34 seconds ==================================================================================================PS C:\Users\username\sources\AI-on-the-edge-device\code>
```

--- a/code/main/main.cpp
+++ b/code/main/main.cpp
@@ -142,13 +142,7 @@ void task_NoSDBlink(void *pvParameter)
 extern "C" void app_main(void)
 {
     TickType_t xDelay;
-
-
-    printf("Do Reset Camera\n");
-    PowerResetCamera();
-    Camera.InitCam();
-    Camera.LightOnOff(false); 
-
+ 
     if (!Init_NVS_SDCard())
     {
         xTaskCreate(&task_NoSDBlink, "task_NoSDBlink", configMINIMAL_STACK_SIZE * 64, NULL, tskIDLE_PRIORITY+1, NULL);
@@ -176,8 +170,8 @@ extern "C" void app_main(void)
        printf("DNS IP: %s\n", dns);
 
 
-    wifi_init_sta(ssid, passwd, hostname, ip, gateway, netmask, dns);
-    
+    wifi_init_sta(ssid, passwd, hostname, ip, gateway, netmask, dns);   
+
 
     xDelay = 2000 / portTICK_PERIOD_MS;
     printf("main: sleep for : %ldms\n", (long) xDelay);
@@ -198,8 +192,8 @@ extern "C" void app_main(void)
     printf("time %s\n", zw.c_str());    
 
 //    Camera.InitCam();
-//    Camera.LightOnOff(false); 
-    xDelay = 2000 / portTICK_PERIOD_MS;
+//    Camera.LightOnOff(false);
+     xDelay = 2000 / portTICK_PERIOD_MS;
     printf("main: sleep for : %ldms\n", (long) xDelay);
     vTaskDelay( xDelay ); 
 
@@ -215,6 +209,22 @@ extern "C" void app_main(void)
     register_server_main_uri(server, "/sdcard");
 
     printf("vor dotautostart\n");
-    TFliteDoAutoStart();
+
+    // init camera module
+    printf("Do Reset Camera\n");
+    PowerResetCamera();
+    esp_err_t cam = Camera.InitCam();
+    if (cam != ESP_OK) {
+            ESP_LOGE(TAGMAIN, "Failed to initialize camera module. "
+                "Check that your camera module is working and connected properly.");
+
+            LogFile.SwitchOnOff(true);
+            LogFile.WriteToFile("Failed to initialize camera module. "
+                    "Check that your camera module is working and connected properly.");
+            LogFile.SwitchOnOff(false);
+    } else {
+        Camera.LightOnOff(false);
+        TFliteDoAutoStart();
+    }
 }
 

